packages/components/src/higher-order/with-filters/index.js

```diff
@@ -24,45 +24,82 @@ const ANIMATION_FRAME_PERIOD = 16;
+		 * avoid excessive calls to `applyFilters` when many instances exist.
+		 *
+		 * @type {Component}
+		 */
+		let FilteredComponent = applyFilters( hookName, OriginalComponent );
+
+		class FilteredComponentRenderer extends Component {
 			constructor( props ) {
 				super( props );
 
-				this.onHooksUpdated = this.onHooksUpdated.bind( this );
-				this.Component = applyFilters( hookName, OriginalComponent );
-				this.namespace = uniqueId( 'core/with-filters/component-' );
-				this.throttledForceUpdate = debounce( () => {
-					this.Component = applyFilters( hookName, OriginalComponent );
-					this.forceUpdate();
-				}, ANIMATION_FRAME_PERIOD );
+				this.throttledForceUpdate = debounce(
+					() => this.forceUpdate(),
+					ANIMATION_FRAME_PERIOD
+				);
+			}
+
+			componentDidMount() {
+				FilteredComponentRenderer.instances.push( this );
 
-				addAction( 'hookRemoved', this.namespace, this.onHooksUpdated );
-				addAction( 'hookAdded', this.namespace, this.onHooksUpdated );
+				// If there were previously no mounted instances for components
+				// filtered on this hook, add the hook handler.
+				if ( FilteredComponentRenderer.instances.length === 1 ) {
+					addAction( 'hookRemoved', namespace, onHooksUpdated );
+					addAction( 'hookAdded', namespace, onHooksUpdated );
+				}
 			}
 
-			/** @inheritdoc */
 			componentWillUnmount() {
 				this.throttledForceUpdate.cancel();
-				removeAction( 'hookRemoved', this.namespace );
-				removeAction( 'hookAdded', this.namespace );
-			}
 
-			/**
-			 * When a filter is added or removed for the matching hook name, the wrapped component should re-render.
-			 *
-			 * @param {string} updatedHookName  Name of the hook that was updated.
-			 */
-			onHooksUpdated( updatedHookName ) {
-				if ( updatedHookName === hookName ) {
-					this.throttledForceUpdate();
+				FilteredComponentRenderer.instances = without(
+					FilteredComponentRenderer.instances,
+					this
+				);
+
+				// If this was the last of the mounted components filtered on
+				// this hook, remove the hook handler.
+				if ( FilteredComponentRenderer.instances.length === 0 ) {
+					removeAction( 'hookRemoved', namespace );
+					removeAction( 'hookAdded', namespace );
 				}
 			}
 
-			/** @inheritdoc */
 			render() {
-				return <this.Component { ...this.props } />;
+				return <FilteredComponent { ...this.props } />;
 			}
-		};
+		}
+
+		FilteredComponentRenderer.instances = [];
+
+		/**
+		 * When a filter is added or removed for the matching hook name, each
+		 * mounted instance should re-render with the new filters having been
+		 * applied to the original component.
+		 *
+		 * @param {string} updatedHookName Name of the hook that was updated.
+		 */
+		function onHooksUpdated( updatedHookName ) {
+			if ( updatedHookName !== hookName ) {
+				return;
+			}
+
+			// Recreate the filtered component.
+			FilteredComponent = applyFilters( hookName, OriginalComponent );
```

Comment:
Moving this back to  `throttledForceUpdate` would mean each component re-evaluates the `applyFilters` itself, rather than doing it just once for the shared definition. This is what I meant by "requiring the refactor" in https://github.com/WordPress/gutenberg/pull/13231#discussion_r246045044, since if we handled the debounce outside the components, it would only call `applyFilters` once, after the delay.